### PR TITLE
Add a custom eslint rule to check the constructors

### PR DESCRIPTION
**Pull request by jamie-lynch**
## What does this change?

This PR creates a custom ESLint rule which is used to check that the constructors of all constructs and patterns follow the best practice laid out in the [ADR](https://github.com/guardian/cdk/blob/main/docs/architecture-decision-records/002-component-constuctors.md).

Any constructs breaking those rules have been updated. In all cases except one, this change is non-breaking.

The InstanceRole pattern has had an `id` parameter added which means that any stacks implementing it will need to be updated.

## Does this change require changes to existing projects or CDK CLI?

Yes! The InstanceRole pattern has now had an `id` parameter added to conform to the style guide. This means that any stacks implementing it will need to be updated.

## How to test

Run the lint command to see if any errors are detected. Check a deliberate violation to see it catch things. 

## How can we measure success?

The best practice agreed around component constructors is verified by an automated check rather than relying on peer reviews to catch any violations.

## Have we considered potential risks?

The rule will either fail to catch violations or detect numerous false positives.

### CODE REVIEW

File: eslint/rules/valid-constructors.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+          // 7. Third parameter (if exists) must called props
+          if (getStandardOrLeftProp(props, "name") !== "props") {
+            return context.report(
+              node,
+              props.loc,
+              `The third parameter in a construct or pattern contructor must be called props`
+            );
+          }
+
+          // 10. Third parameter type should be custom
```
Comment (jacobwinch) [inline eslint/rules/valid-constructors.js:124]: Minor: should we swap the order of 8 & 10?

**Review by jacobwinch — APPROVED** [review]
Disclaimer: I've never written a custom eslint rule before, but based on the docs and a quick local test this looks good to me 👍

File: eslint/rules/valid-constructors.js (hunk 1 above)
Comment (jamie-lynch) [reply eslint/rules/valid-constructors.js]: Good point!

**Comment by akash1810** [comment]
Oooh making the linter work for us, this is great!

**Pull request merged**